### PR TITLE
langref: add example for errdefer |err| sytnax

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

File: doc/langref/test_errdefer_capture.zig [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        captured.* = err;
+    }
+    return error.GeneralFailure;
+}
+
+test "errdefer capture" {
+    var captured: ?anyerror = null;
+
+    if (captureError(&captured)) unreachable else |err| {
+        try std.testing.expectEqual(error.GeneralFailure, captured.?);
```
Comment (matklad) [inline doc/langref/test_errdefer_capture.zig:14]: :warning: if I remove `.?` and replace `captured.?` with just `captured` the test fails with `expected error.GeneralFailure, found error.GeneralFailure`. Seems like a bug in the language: either this code should not compile without `?`, or it should return true.

Comment (rohlem) [reply doc/langref/test_errdefer_capture.zig:14]: Could just be that std.testing.expectEqual doesn't look at the payload of optionals, but when printing the output is the same.

Comment (Rexicon226) [reply doc/langref/test_errdefer_capture.zig:14]: What's happening is that the "found" `error.GeneralFailure`, is an instant of the `anyerror`, whereas the "expected" `error.GeneralFailure` is part of the IES. There is no well defined behaviour for coercing an error into an `error{X}!?anyerror`. Does the error go into the error set or into the `anyerror`? Here it's just become two different things that doesn't equal.

Comment (rohlem) [reply doc/langref/test_errdefer_capture.zig:14]: @Rexicon226 I do think that there is well-defined behavior: Anything that isn't an undecorated error is a payload; that's the way type resolution for `return` works (which is the same mechanism behind multi-argument `@TypeOf`).
The reason one `?anyerror` is a payload I think is that it's decorated by `?`, I don't think it being part of an inferred error set matters.

Comment (rohlem) [reply doc/langref/test_errdefer_capture.zig:14]: created #20442 for discussion

Comment (matklad) [reply doc/langref/test_errdefer_capture.zig:14]: Ah, got it! Just to restate it, the problem here is that expectEquals tries to find a common type between its arguments. The lhs type here is `error{GeneralFailure}`, the rhs type is `?anyerror`. And there are three "reasonable" answers as to what's the common type here:

1) common type is `?anyerror` -> convert singleton error set to anyerror, that convert that to an optional. 
2) common type is an error union `error{GeneralFailure} ! ?anyerror` -> using the general rule that the common type for `error{...}` and `T` is `error{...}!T`
3) no common type, as this is ambigious for the human. 

Currently, Zig uses the option 2.

Comment (Rexicon226) [reply doc/langref/test_errdefer_capture.zig:14]: @rohlem I see, you’re right. Didn’t fully consider how it was working. Thanks for the issue :)

**Comment by andrewrk** [comment]
Thanks!

**Pull request merged**